pkg/controller/build/helpers_test.go

```diff
@@ -80,3 +82,119 @@ func TestValidateOnClusterBuildConfig(t *testing.T) {
 		})
 	}
 }
+
+// This test validates that we have correctly identified if the MachineOSBuild
+// should be updated based upon comparing the old and current status of the
+// MachineOSBuild. It is worth noting that the current MachineOSBuild status
+// can come from the imagebuilder.MachineOSBuildStatus() method which maps the
+// current job state to the MachineOSBuild state.
+func TestIsMachineOSBuildStatusUpdateNeeded(t *testing.T) {
+	t.Parallel()
+
+	initialConditions := func() map[mcfgv1.BuildProgress][]metav1.Condition {
+		return map[mcfgv1.BuildProgress][]metav1.Condition{
+			// This value is not part of the OCL API and is here solely for testing purposes.
+			"Initial": apihelpers.MachineOSBuildInitialConditions(),
+		}
+	}
+
+	testCases := []struct {
+		name     string
+		old      map[mcfgv1.BuildProgress][]metav1.Condition
+		current  map[mcfgv1.BuildProgress][]metav1.Condition
+		expected bool
+	}{
+		// These are valid state transitions. In other words, when one of these
+		// state transitions is identified, the MachineOSBuild status object should
+		// be updated.
+		{
+			name:     "Initial -> Terminal",
+			old:      initialConditions(),
+			current:  ctrlcommon.MachineOSBuildTerminalStates(),
+			expected: true,
+		},
+		{
+			name:     "Initial -> Transient",
+			old:      initialConditions(),
+			current:  ctrlcommon.MachineOSBuildTransientStates(),
+			expected: true,
+		},
+		{
+			name:     "Transient -> Terminal",
+			old:      ctrlcommon.MachineOSBuildTransientStates(),
+			current:  ctrlcommon.MachineOSBuildTerminalStates(),
+			expected: true,
+		},
+		{
+			name: "Pending -> Running",
+			old: map[mcfgv1.BuildProgress][]metav1.Condition{
+				mcfgv1.MachineOSBuildPrepared: ctrlcommon.MachineOSBuildTransientStates()[mcfgv1.MachineOSBuildPrepared],
+			},
+			current: map[mcfgv1.BuildProgress][]metav1.Condition{
+				mcfgv1.MachineOSBuilding: ctrlcommon.MachineOSBuildTransientStates()[mcfgv1.MachineOSBuilding],
+			},
+			expected: true,
+		},
+		// These are invalid state transitions. In other words, when one of these
+		// state transitions is observed, the MachineOSBuild object should not be
+		// updated because they are invalid and make no sense.
+		{
+			name:     "Terminal -> Initial",
```

Comment:
Question: For the case where a build fails, we fix the issue, and then trigger the build again with the rebuild annotation, having this restriction shouldn't cause an issue right?

The MOSB is deleted and recreated for this case but it is created with the same name. I don't think trying to update the state for the new MOSB with the same name should cause an issue but I think it is something worth confirming.

Reply:
It shouldn't in that because it is a new MOSB.